pkg/utils/types/time_series.go

```diff
@@ -0,0 +1,83 @@
+package types
+
+import (
+	"sync"
+)
+
+// Timestamp is a regulation of time types
+type Timestamp int
+
+// State represent a change of value - the new value and the time of the change
+type State[T any] struct {
+	StartTime Timestamp
+	Val       T
+}
+
+// TimeSeries represents a value that changes over time.
+// It exposes the value at a given time, assuming that between two changes the value is constant.
+// The object is designed to be thread-safe.
+type TimeSeries[T any] struct {
+	states []State[T]
+	mu     sync.RWMutex
+}
+
+func NewTimeSeries[T any](defaultVal T) *TimeSeries[T] {
+	defaultState := State[T]{
+		StartTime: 0,
+		Val:       defaultVal,
+	}
+	return &TimeSeries[T]{
+		states: []State[T]{defaultState},
+	}
+}
+
+// AddState add a change in the value - from given time until next change, the value is the new one.
+func (co *TimeSeries[T]) AddState(newState State[T]) {
+	co.mu.Lock()
+	defer co.mu.Unlock()
+	insertIndex := co.getRelevantStateIndex(newState.StartTime) + 1
+	stateList := make([]State[T], len(co.states)+1)
```

Comment:
TODO: This might not be very efficient if called frequently. We could use a linked list OR, at least, allocate a bigger slice and only creating a new only when needed.

Reply:
> Shouldn't this be []*State[T] ?

I thought no, because the `val` can be just a reference so doing double dereference seems weird to me.
However, if the timestamp will be developed to be a more unique time then you are right.

> TODO: This might not be very efficient if called frequently. We could use a linked list OR, at least, allocate a bigger slice and only creating a new only when needed.

However, you will pay for having a linked list by having inefficient search. You won't be able to do binary search in that case for example.
You might be right that we should allocate bigger sized array, I need to think about it.

Reply:
> I thought no, because the val can be just a reference so doing double dereference seems weird to me. However, if the timestamp will be developed to be a more unique time then you are right.

If we have:

```go
type State[T any] struct {
    StartTime Timestamp
    Val *T
}
```
we would enforce Val to always be a pointer (or get a compile-time error). This way you would guarantee no shallow copies are done (which is likely the case you need as Val is always a pointer to some instance).

Reply:
> However, you will pay for having a linked list by having inefficient search. 

Sure, a binary tree then (considering StartTime can always be sorted and used as the node's keys). Might be faster than creating slices over and over (or having to manage a bigger one) and will allow faster searches (search will continue being binary and same or better complexity).

Reply:
I agree with you that binary tree is better for the long term usage of this object.
However, currently I don't anticipate that I will need more than 10 states in an object.
Thus, the current implementation is probably better in performance than binary tree or AVL tree.
The current exposed API will support change in the implementation in the future, so I don't see a specific reason to change the implementation right now.
I will add a TODO for it in the code, as in the long term when other places will adopt this type it will increase performance.